src/qat/backend/analysis_passes.py

```diff
@@ -0,0 +1,409 @@
+                        for s, inst in enumerate(builder.instructions[h + 1 :])
+                        if isinstance(inst, (Sweep, Repeat, EndSweep, EndRepeat))
+                    ),
+                    None,
+                )
+                if k:
+                    next_headers.add(k + h + 1)
+                    next_flow.add((h, k + h + 1))
+                    src.elements.extend(range(src.tail() + 1, k + h + 1))
+                    dest = cfg.get_or_create_node(k + h + 1)
+                    cfg.get_or_create_edge(src, dest)
+
+        if next_headers == set(headers) and next_flow == set(flow):
+            # TODO - In-place use instructions instead of indices directly
+            for n in cfg.nodes:
+                n.elements = [builder.instructions[i] for i in n.elements]
+            return
+
+        self._build_cfg(builder, cfg)
+
+
+class LegalisationPass(AnalysisPass):
+    """
+    The sole purpose of this pass is to allow us to understand what can and what cannot be run on the control stack.
+
+    An instruction is legal if it has a direct equivalent in the programming model implemented by the control stack.
+    The notion of "legal" is highly determined by the hardware features of the control stack as well as its
+    programming model.
+
+    Most control stacks such as Qblox have a direct ISA-level representation for basic quantum instructions such as
+    frequency and phase manipulation instructions, arithmetic instructions such as add, branching instructions
+    such as jump.
+
+    Particularly:
+    1) A sweep instruction is illegal because it specifies unclear iteration semantics.
+    2) A repeat instruction with a very high repetition count is illegal because acquisition memory on a sequencer
+    is limited. This requires optimal batching of the repeat instruction into maximally supported batches of smaller
+    repeat counts.
+    3) Device updates/assigns in general are illegal because they are bound to a sweep instruction via a variable
+    name. In fact, a sweep variable remains obscure and opaque until a device update bound to it is encountered
+    to reveal what it intends to do with the variable (usually on the instruction builder or on the hardware model).
+    We say that a device update carries meaning for the sweep variable and materialises its intention.
+
+    A brute force implementation would increase the depth of loop nests via repeat batching as well as bubble
+    outwards any illegal instructions.
+    """
+
+    def run(self, builder: InstructionBuilder, res_mgr: ResultManager, *args, **kwargs):
+        pass
+
+
+class LifetimePass(AnalysisPass):
+    """
+    The end goal of this pass is to facilitate sequencer allocation on the control hardware. Much like classical
+    register allocation techniques, this pass is intended to perform "channel liveness" analysis.
+
+    A logical channel is alive at some point T1 in the builder (analogically to a classical program) if it is
```

Comment:
Don't use T1 and T2! Cause they also mean qubit life time!!

Reply:
hahahaa, oh yes I forgot they mean sth elsewhere, will rename 😄